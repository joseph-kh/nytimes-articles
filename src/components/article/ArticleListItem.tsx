import React from "react";
import { View, Image, StyleSheet, TouchableOpacity } from "react-native";
import { MaterialIcons } from "@expo/vector-icons";
import { horizontalScale, moderateScale, verticalScale } from "@/utils/scaler";
import Row from "@/components/ui/containers/Row";
import RegularText from "@/components/ui/texts/RegularText";
import Title from "@/components/article/Title";
import Author from "@/components/article/Author";
import PublishedDate from "@/components/article/PublishedDate";
import COLORS from "@/theme/colors";

type ArticleListItemProps = {
  article: Article;
  onArticlePress?: () => void;
};

const ArticleListItem: React.FC<ArticleListItemProps> = (props) => {
  const { onArticlePress, article } = props;

  const { title, published_date, section, byline, media } = article;

  return (
    <TouchableOpacity onPress={onArticlePress} style={styles.container}>
      <Image source={{ uri: media }} style={styles.imageStyle} />

      <View style={styles.infoContainer}>
        <Title title={title} style={styles.titleStyle} />
        <Author author={byline} />
        <Row style={{ justifyContent: "space-between" }}>
          <RegularText style={styles.grayText}>{section}</RegularText>
          <PublishedDate publishedDate={published_date} />
        </Row>
      </View>

      <View style={styles.iconContainer}>
        <MaterialIcons
          name="arrow-forward-ios"
          size={moderateScale(26)}
          color={COLORS.gray}
        />
      </View>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    alignItems: "center",
    paddingVertical: verticalScale(10),
    paddingHorizontal: horizontalScale(15),
  },

  infoContainer: {
    flex: 0.8,
    marginLeft: horizontalScale(10),
  },

  titleStyle: {
    marginBottom: verticalScale(15),
  },

  grayText: {
    color: COLORS.gray,
  },

  imageStyle: {
    width: moderateScale(50),
    height: moderateScale(50),
    borderRadius: moderateScale(50) / 2,
  },

  iconContainer: {
    alignItems: "flex-end",
    flex: 0.2,
  },
});

export default ArticleListItem;
