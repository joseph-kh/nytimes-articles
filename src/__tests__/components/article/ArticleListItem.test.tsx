import React from "react";
import renderer from "react-test-renderer";
import { TouchableOpacity } from "react-native";
import ArticleListItem from "@/components/article/ArticleListItem";

const mockArticle = {
  id: "1",
  title: "Sample Article",
  published_date: "2023-09-01",
  section: "React Native",
  byline: "John Doe",
  media:
    "https://cdn.discordapp.com/attachments/1148605640856572054/1148605814857277581/56117400-9a911800-5f85-11e9-878b-3f998609a6c8.jpg",
  abstract: "samle abstract",
};

describe("ArticleListItem Component", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(<ArticleListItem article={mockArticle} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("calls onArticlePress when pressed", () => {
    const mockOnPress = jest.fn();
    const component = renderer.create(
      <ArticleListItem article={mockArticle} onArticlePress={mockOnPress} />
    );

    const touchableOpacity = component.root.findByType(TouchableOpacity);
    touchableOpacity.props.onPress();

    expect(mockOnPress).toHaveBeenCalled();
  });
});
