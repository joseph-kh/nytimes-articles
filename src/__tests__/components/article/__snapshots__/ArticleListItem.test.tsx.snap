// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleListItem Component renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "opacity": 1,
      "paddingHorizontal": 30,
      "paddingVertical": 16.428571428571427,
    }
  }
>
  <Image
    source={
      {
        "uri": "https://cdn.discordapp.com/attachments/1148605640856572054/1148605814857277581/56117400-9a911800-5f85-11e9-878b-3f998609a6c8.jpg",
      }
    }
    style={
      {
        "borderRadius": 37.5,
        "height": 75,
        "width": 75,
      }
    }
  />
  <View
    style={
      {
        "flex": 0.8,
        "marginLeft": 20,
      }
    }
  >
    <Text
      style={
        {
          "fontFamily": "Bold",
          "fontSize": 21,
          "marginBottom": 24.642857142857142,
        }
      }
      title="Sample Article"
    >
      Sample Article
    </Text>
    <Text
      author="John Doe"
      style={
        {
          "color": "#7e7e7e",
          "fontFamily": "Italic",
          "fontSize": 21,
        }
      }
    >
      John Doe
    </Text>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <Text
        style={
          {
            "color": "#7e7e7e",
            "fontFamily": "Regular",
            "fontSize": 21,
          }
        }
      >
        React Native
      </Text>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Text />
        <Text
          style={
            {
              "color": "#7e7e7e",
              "fontFamily": "Regular",
              "fontSize": 21,
            }
          }
        >
          2023-09-01
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "flex-end",
        "flex": 0.2,
      }
    }
  >
    <Text />
  </View>
</View>
`;
